File: codility/go/9-max_slice.go
package golang

// https://app.codility.com/programmers/lessons/9-maximum_slice_problem/max_profit/
func maxProfit(A []int) int {
	if len(A) == 0 {
		return 0
	}
	if len(A) == 1 {
		return 0
	}
	if len(A) == 2 {
		if A[1] > A[0] {
			return A[1] - A[0]
		}
		return 0
	}
	profit, buy, sell := 0, A[0], A[0]
	for day := 1; day < len(A); day++ {
		price := A[day]
		// if price lower than buy price, set both buy and sell price to this day
		// if price higher than sell price, set sell price to today
		// what is the potential profit? update if higher than best so far
		if price < buy {
			buy = price
			sell = price
			continue
		}

		if price > sell {
			sell = price

			if sell-buy > profit {
				profit = sell - buy
			}
		}
	}

	return profit
}

// https://app.codility.com/programmers/lessons/9-maximum_slice_problem/max_slice_sum/
func maxSliceSum(A []int) int {
	if len(A) == 0 {
		return 0
	}
	maxSum, currentSum := A[0], A[0]
	for i := 1; i < len(A); i++ {
		currentSum += A[i]
		if A[i] > currentSum {
			currentSum = A[i]
		}
		if currentSum > maxSum {
			maxSum = currentSum
		}
	}

	return maxSum
}

// https://app.codility.com/programmers/lessons/9-maximum_slice_problem/max_double_slice_sum/
func maxDoubleSliceSum(A []int) int {
	// leftToRight, rightToLeft slices
	// save slice sum at each position
	ltr, rtl := make([]int, len(A)), make([]int, len(A))
	for i := 1; i < len(A)-1; i++ {
		// ltr
		// add current value to previous value, or reset to zero
		ltr[i] = ltr[i-1] + A[i]
		if ltr[i] < 0 {
			ltr[i] = 0
		}

		// rtl
		// ii mirrors i
		ii := len(A) - 1 - i
		rtl[ii] = rtl[ii+1] + A[ii]
		if rtl[ii] < 0 {
			rtl[ii] = 0
		}

		// [MOVED for clarity] When half of the slice passed, check for best total sum
		//if i >= len(A)/2 {
		//	// ltr and rtl for both sides of i
		//	candidate := ltr[i-1] + rtl[i+1]
		//	if candidate > best {
		//		best = candidate
		//	}
		//
		//	// ltr and rtl for both sides of ii
		//	candidate = ltr[ii+1] + rtl[ii-1]
		//	if candidate > best {
		//		best = candidate
		//	}
		//}
	}

	// run through indexes again, for each, check both slices for best sum
	// this could also be done in the initial loop after i has passed half of the values
	var maxSum int
	for i := 1; i < len(A)-1; i++ {
		candidate := ltr[i-1] + rtl[i+1]
		if candidate > maxSum {
			maxSum = candidate
		}
	}

	return maxSum
}

File: codility/py/9-max_slice.py
def example(A):
    current = best = 0
    for a in A:
        current = max(0,current+a) # If the sum is negative, reset to zero instead
        best = max(current, best)
    return best


assert example([3, 2, -6, 4, 0]) == 5
assert example([-2, -3, -1]) == 0
assert example([-2, -3, -1, 4]) == 4
assert example([1, 2, 3]) == 6
assert example([1, 2, -3]) == 3

def max_profit(A):
    if not A or len(A) < 2:
        return 0

    min_price, best_profit = A[0], 0
    for day in range(1,len(A)):
        price = A[day]
        best_profit = max(best_profit, price - min_price)
        min_price = min(min_price, price) # Update for future sells
    return best_profit

assert max_profit([23171, 21011, 21123, 21366, 21013, 21367]) == 356
assert max_profit([1, 2, 3, 4, 5]) == 4
assert max_profit([5, 4, 3, 2, 1]) == 0
assert max_profit([5, 1, 2, 3, 4]) == 3
assert max_profit([1, 2, 3, 2, 1]) == 2
assert max_profit([7, 1, 5, 3, 6, 4]) == 5
assert max_profit([7, 6, 4, 3, 1]) == 0

def max_slice_sum(A):
    if not A:
        return 0
    current = best = A[0]
    for i in range(1,len(A)):
        current = max(A[i],current+A[i])
        best = max(best,current)
    return best

assert max_slice_sum([3, 2, -6, 4, 0]) == 5
assert max_slice_sum([-2, -3, -1]) == -1
assert max_slice_sum([-2, -3, -1, 4]) == 4
assert max_slice_sum([1, 2, 3]) == 6
assert max_slice_sum([1, 2, -3]) == 3
